vllm/model_executor/models/h2ovl.py

```diff
@@ -0,0 +1,400 @@
+
+    def __init__(self):
+        super().__init__(IMG_START, IMG_END, IMG_CONTEXT)
+
+    def input_processor(
+        self,
+        ctx: InputContext,
+        inputs: DecoderOnlyInputs,
+        *,
+        max_dynamic_patch: Optional[int] = None,
+    ) -> DecoderOnlyInputs:
+        multi_modal_data = inputs.get("multi_modal_data")
+        if multi_modal_data is None or "image" not in multi_modal_data:
+            return inputs
+
+        model_config = ctx.model_config
+        hf_config = ctx.get_hf_config()
+
+        image_data = multi_modal_data["image"]
+        num_patches = get_internvl_num_patches(hf_config)
+
+        # can only get the total blocks num after the image fully processed
+        num_blocks_calculator = image_to_pixel_values_wrapper(
+            hf_config, max_dynamic_patch=max_dynamic_patch)
+
+        if isinstance(image_data, Image.Image):
+            num_blocks = num_blocks_calculator(image_data).shape[0]
+            image_feature_sizes = [num_blocks * num_patches]
+
+        elif is_list_of(image_data, Image.Image):
+            # Do not use MSAC for multi images
+            hf_config.use_msac = False
+            image_feature_sizes = []
+            for image in image_data:
+                num_blocks = num_blocks_calculator(image).shape[0]
+                image_feature_sizes.append(num_blocks * num_patches)
+
+        elif isinstance(image_data, torch.Tensor):
+            num_images, image_feature_size, hidden_size = image_data.shape
+            image_feature_sizes = [image_feature_size]
+        else:
+            raise TypeError(f"Invalid image type: {type(image_data)}")
+
+        tokenizer = cached_get_tokenizer(
+            model_config.tokenizer,
+            trust_remote_code=model_config.trust_remote_code,
+        )
+
+        prompt = inputs.get("prompt")
+        prompt_token_ids = inputs["prompt_token_ids"]
+        if prompt is None:
+            prompt = tokenizer.decode(prompt_token_ids)
+
+        new_prompt = self._expand_image_prompt(prompt, image_feature_sizes,
+                                               num_patches)
+        new_prompt_token_ids = tokenizer.encode(new_prompt)
+
+        return token_inputs(
+            prompt=prompt,
+            prompt_token_ids=new_prompt_token_ids,
+            multi_modal_data=multi_modal_data,
+        )
+
+    def input_mapper(
+        self,
+        ctx: InputContext,
+        data: object,
+        *,
+        max_dynamic_patch: Optional[int] = None,
+    ):
+        hf_config = ctx.get_hf_config()
+
+        image_pixel_values_mapper = image_to_pixel_values_wrapper(
+            hf_config, max_dynamic_patch)
+
+        if isinstance(data, Image.Image):
+            data = image_pixel_values_mapper(data)
+            data = data.unsqueeze(0)
+        elif is_list_of(data, Image.Image):
+            hf_config.use_msac = False
+            data = [image_pixel_values_mapper(img) for img in data]
```

Comment:
Per my comment above - this second call of image processing is what we should avoid